File: tests/mcp/tools/definitions/foreshadowing_view_test.ts
/**
 * foreshadowing_view MCPツールテスト（TDD Red Phase）
 */

import { assertEquals, assertExists } from "@std/assert";
import { foreshadowingViewTool } from "@storyteller/mcp/tools/definitions/foreshadowing_view.ts";
import type { Foreshadowing } from "@storyteller/types/v2/foreshadowing.ts";

Deno.test("foreshadowing_view MCPツール", async (t) => {
  await t.step("ツール名がforeshadowing_viewであること", () => {
    assertEquals(foreshadowingViewTool.name, "foreshadowing_view");
  });

  await t.step("descriptionが設定されていること", () => {
    assertExists(foreshadowingViewTool.description);
  });

  await t.step("inputSchemaにidとlistプロパティがあること", () => {
    const properties = foreshadowingViewTool.inputSchema.properties;
    assertExists(properties);
    assertExists(properties.id);
    assertExists(properties.list);
    assertExists(properties.status);
  });

  await t.step("list=trueで伏線一覧を取得できること", async () => {
    const tempDir = await Deno.makeTempDir();

    try {
      const foreshadowingsDir = `${tempDir}/src/foreshadowings`;
      await Deno.mkdir(foreshadowingsDir, { recursive: true });

      const foreshadowing: Foreshadowing = {
        id: "sword",
        name: "古びた剣",
        type: "chekhov",
        summary: "床板の下から発見される剣",
        planting: {
          chapter: "chapter_01",
          description: "剣を発見する",
        },
        status: "planted",
      };

      await Deno.writeTextFile(
        `${foreshadowingsDir}/sword.ts`,
        `import type { Foreshadowing } from "@storyteller/types/v2/foreshadowing.ts";
export const sword: Foreshadowing = ${JSON.stringify(foreshadowing, null, 2)};`,
      );

      const result = await foreshadowingViewTool.execute(
        { list: true },
        { projectRoot: tempDir },
      );

      assertEquals(result.isError, false);
      assertExists(result.content);
      assertEquals(result.content.length > 0, true);
    } finally {
      await Deno.remove(tempDir, { recursive: true });
    }
  });

  await t.step("idで特定の伏線を取得できること", async () => {
    const tempDir = await Deno.makeTempDir();

    try {
      const foreshadowingsDir = `${tempDir}/src/foreshadowings`;
      await Deno.mkdir(foreshadowingsDir, { recursive: true });

      const foreshadowing: Foreshadowing = {
        id: "prophecy",
        name: "予言",
        type: "prophecy",
        summary: "勇者の予言",
        planting: {
          chapter: "chapter_02",
          description: "予言が告げられる",
        },
        status: "planted",
      };

      await Deno.writeTextFile(
        `${foreshadowingsDir}/prophecy.ts`,
        `import type { Foreshadowing } from "@storyteller/types/v2/foreshadowing.ts";
export const prophecy: Foreshadowing = ${
          JSON.stringify(foreshadowing, null, 2)
        };`,
      );

      const result = await foreshadowingViewTool.execute(
        { id: "prophecy" },
        { projectRoot: tempDir },
      );

      assertEquals(result.isError, false);
      assertExists(result.content);
    } finally {
      await Deno.remove(tempDir, { recursive: true });
    }
  });

  await t.step("statusフィルタが動作すること", async () => {
    const tempDir = await Deno.makeTempDir();

    try {
      const foreshadowingsDir = `${tempDir}/src/foreshadowings`;
      await Deno.mkdir(foreshadowingsDir, { recursive: true });

      const planted: Foreshadowing = {
        id: "planted_one",
        name: "未回収伏線",
        type: "hint",
        summary: "まだ回収されていない",
        planting: {
          chapter: "chapter_01",
          description: "設置",
        },
        status: "planted",
      };

      const resolved: Foreshadowing = {
        id: "resolved_one",
        name: "回収済み伏線",
        type: "mystery",
        summary: "すでに回収済み",
        planting: {
          chapter: "chapter_01",
          description: "設置",
        },
        status: "resolved",
        resolutions: [
          {
            chapter: "chapter_05",
            description: "回収",
            completeness: 1.0,
          },
        ],
      };

      await Deno.writeTextFile(
        `${foreshadowingsDir}/planted_one.ts`,
        `import type { Foreshadowing } from "@storyteller/types/v2/foreshadowing.ts";
export const planted_one: Foreshadowing = ${JSON.stringify(planted, null, 2)};`,
      );

      await Deno.writeTextFile(
        `${foreshadowingsDir}/resolved_one.ts`,
        `import type { Foreshadowing } from "@storyteller/types/v2/foreshadowing.ts";
export const resolved_one: Foreshadowing = ${
          JSON.stringify(resolved, null, 2)
        };`,
      );

      const result = await foreshadowingViewTool.execute(
        { list: true, status: "planted" },
        { projectRoot: tempDir },
      );

      assertEquals(result.isError, false);
      assertExists(result.content);
      // 結果にplanted_oneのみが含まれることを確認
      const content = result.content[0];
      if (content.type === "text") {
        assertEquals(content.text.includes("planted_one"), true);
        assertEquals(content.text.includes("resolved_one"), false);
      }
    } finally {
      await Deno.remove(tempDir, { recursive: true });
    }
  });

  await t.step("存在しないIDでエラーを返すこと", async () => {
    const tempDir = await Deno.makeTempDir();

    try {
      const foreshadowingsDir = `${tempDir}/src/foreshadowings`;
      await Deno.mkdir(foreshadowingsDir, { recursive: true });

      const result = await foreshadowingViewTool.execute(
        { id: "nonexistent" },
        { projectRoot: tempDir },
      );

      assertEquals(result.isError, true);
    } finally {
      await Deno.remove(tempDir, { recursive: true });
    }
  });
});

File: internal/mcp/tools/foreshadowing_view_test.go
package tools

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestForeshadowingViewTool_Definition(t *testing.T) {
	def := ForeshadowingViewTool{}.Definition()
	if def.Name != "foreshadowing_view" {
		t.Errorf("name = %q, want %q", def.Name, "foreshadowing_view")
	}
	if def.Description == "" {
		t.Error("description should not be empty")
	}
	if len(def.InputSchema) == 0 {
		t.Error("input schema empty")
	}
}

// newForeshadowingProjectWithEntities creates a project dir with a foreshadowing TS fixture.
func newForeshadowingProjectWithEntities(t *testing.T) string {
	t.Helper()
	root := t.TempDir()
	writeFile(t, filepath.Join(root, ".storyteller.json"), `{"version":"1.0.0"}`)
	dir := filepath.Join(root, "src", "foreshadowings")
	if err := os.MkdirAll(dir, 0o755); err != nil {
		t.Fatal(err)
	}
	// Two foreshadowings: one planted, one resolved
	writeFile(t, filepath.Join(dir, "ancient_sword.ts"), `
export const ancient_sword = {
  "id": "ancient_sword",
  "name": "古びた剣",
  "type": "chekhov",
  "summary": "床板の下に隠された剣",
  "planting": {
    "chapter": "chapter_01",
    "description": "床板の下から発見"
  },
  "status": "planted"
};
`)
	writeFile(t, filepath.Join(dir, "prophecy.ts"), `
export const prophecy = {
  "id": "prophecy",
  "name": "古の予言",
  "type": "prophecy",
  "summary": "神殿に刻まれた予言",
  "planting": {
    "chapter": "chapter_02",
    "description": "予言が告げられる"
  },
  "status": "resolved"
};
`)
	return root
}

func TestForeshadowingViewTool_Handle_List(t *testing.T) {
	root := newForeshadowingProjectWithEntities(t)
	ec := ExecutionContext{ProjectRoot: root}

	res, err := ForeshadowingViewTool{}.Handle(context.Background(), json.RawMessage(`{}`), ec)
	if err != nil {
		t.Fatalf("Handle returned error: %v", err)
	}
	if res.IsError {
		t.Fatalf("unexpected IsError: %s", res.Content[0].Text)
	}
	text := res.Content[0].Text
	if !strings.Contains(text, "ancient_sword") {
		t.Errorf("expected 'ancient_sword' in list output, got %q", text)
	}
	if !strings.Contains(text, "prophecy") {
		t.Errorf("expected 'prophecy' in list output, got %q", text)
	}
}

func TestForeshadowingViewTool_Handle_ListFilterStatus(t *testing.T) {
	root := newForeshadowingProjectWithEntities(t)
	ec := ExecutionContext{ProjectRoot: root}

	// Filter to planted only
	res, err := ForeshadowingViewTool{}.Handle(context.Background(), json.RawMessage(`{"status":"planted"}`), ec)
	if err != nil {
		t.Fatalf("Handle returned error: %v", err)
	}
	if res.IsError {
		t.Fatalf("unexpected IsError: %s", res.Content[0].Text)
	}
	text := res.Content[0].Text
	if !strings.Contains(text, "ancient_sword") {
		t.Errorf("expected 'ancient_sword' in planted filter, got %q", text)
	}
	if strings.Contains(text, "prophecy") {
		t.Errorf("expected 'prophecy' to be filtered out from planted results, got %q", text)
	}
}

func TestForeshadowingViewTool_Handle_GetByID(t *testing.T) {
	root := newForeshadowingProjectWithEntities(t)
	ec := ExecutionContext{ProjectRoot: root}

	res, err := ForeshadowingViewTool{}.Handle(context.Background(), json.RawMessage(`{"id":"ancient_sword"}`), ec)
	if err != nil {
		t.Fatalf("Handle returned error: %v", err)
	}
	if res.IsError {
		t.Fatalf("unexpected IsError: %s", res.Content[0].Text)
	}
	text := res.Content[0].Text
	if !strings.Contains(text, "古びた剣") {
		t.Errorf("expected name '古びた剣' in detail output, got %q", text)
	}
	if !strings.Contains(text, "chekhov") {
		t.Errorf("expected type 'chekhov' in detail output, got %q", text)
	}
	if !strings.Contains(text, "planted") {
		t.Errorf("expected status 'planted' in detail output, got %q", text)
	}
}

func TestForeshadowingViewTool_Handle_GetByID_NotFound(t *testing.T) {
	root := newForeshadowingProjectWithEntities(t)
	ec := ExecutionContext{ProjectRoot: root}

	res, err := ForeshadowingViewTool{}.Handle(context.Background(), json.RawMessage(`{"id":"nonexistent"}`), ec)
	if err != nil {
		t.Fatalf("Handle returned error: %v", err)
	}
	if !res.IsError {
		t.Errorf("expected IsError=true for missing id, got: %s", res.Content[0].Text)
	}
}

func TestForeshadowingViewTool_Handle_ProjectLoadError(t *testing.T) {
	ec := ExecutionContext{ProjectRoot: "/nonexistent/path"}

	res, err := ForeshadowingViewTool{}.Handle(context.Background(), json.RawMessage(`{}`), ec)
	if err != nil {
		t.Fatalf("Handle returned error: %v", err)
	}
	if !res.IsError {
		t.Errorf("expected IsError=true when project fails to load")
	}
}

func TestForeshadowingViewTool_Handle_EmptyProject(t *testing.T) {
	root := newForeshadowingProject(t)
	ec := ExecutionContext{ProjectRoot: root}

	res, err := ForeshadowingViewTool{}.Handle(context.Background(), json.RawMessage(`{}`), ec)
	if err != nil {
		t.Fatalf("Handle returned error: %v", err)
	}
	if res.IsError {
		t.Fatalf("unexpected IsError for empty project: %s", res.Content[0].Text)
	}
	// Should still return something (e.g. "0 foreshadowings")
	if len(res.Content) == 0 {
		t.Error("expected content even for empty project")
	}
}
